### PR TITLE
properly support deepcopying and serialization of model weights

### LINKED ISSUE #7099: Weights enums cannot be pickled
### 🐛 Describe the bug

Trying to unplickle pickled weights enum object yields an error inside enum implementation (it cannot match constructed object with existing enum values). Here is reproducible example.

```python
import pickle

from torchvision.models import ResNet50_Weights


def main():
    w = ResNet50_Weights.DEFAULT
    kek = pickle.dumps(w)
    lol = pickle.loads(kek)
    print(lol)


if __name__ == '__main__':
    main()
```

This happens because internally enum tries to match newly constructed enum value with declared values. And this fails because `Weights` dataclass has `transforms` field which is a callable. And if that callable is a `functools.partial`, it is technically a different object, even though it was constructed from the same original function with the same partial arguments. This snippet fixes the error, so you should get the idea for probable fix

```python
import functools
from torchvision.models._api import Weights

def new_eq(self, other):
        if not isinstance(other, type(self)):
            return False
        if self.meta != other.meta or self.url != other.url:
            return False
        if not isinstance(self.transforms, functools.partial) or not isinstance(other.transforms, functools.partial):
            return self.transforms == other.transforms
        return all(getattr(self.transforms, a) == getattr(other.transforms, a) for a in ["func", "args", "keywords", "__dict__"])

Weights.__eq__ = new_eq
```

### Versions

```
Collecting environment information...
PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.0.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: version 3.25.0
Libc version: N/A

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:00:33)  [Clang 13.0.1 ] (64-bit runtime)
Python platform: macOS-13.0.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy==0.971
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.4
[pip3] torch==1.12.1
[pip3] torchvision==0.13.1
[conda] numpy                     1.22.4           py39h7df2422_0    conda-forge
[conda] torch                     1.12.1                   pypi_0    pypi
[conda] torchvision               0.13.1                   pypi_0    pypi
```

### PR DESCRIPTION
**Pull request by pmeier**
Fixes #7099 and properly addresses #6871.

The whole idea behind the `TransformsFactory` class is to reimplement `functools.partial`, but with a deepcopy / serialize behavior that better suits our needs. As OP in in #7099 describes, `functools.partial` can be deepcopied and serialized, but the result is no longer equal to input:

```py
import pickle
from copy import deepcopy
from functools import partial


def foo():
    pass


bar = partial(foo)
assert bar != deepcopy(bar)
assert bar != pickle.loads(pickle.dumps(bar))
```

However, we need this equality to work, since it is an integral part of enums. Creating a fresh enum through either of these ops, internally invokes [`Enum.__new__`](https://github.com/python/cpython/blob/a6b889e99d5a8cc63924efacda9e4d4eded4b2fc/Lib/enum.py#L531). In there, a bunch of equality checks are performed against the known members of the enum, i.e. [here](https://github.com/python/cpython/blob/a6b889e99d5a8cc63924efacda9e4d4eded4b2fc/Lib/enum.py#L541) and [here](https://github.com/python/cpython/blob/a6b889e99d5a8cc63924efacda9e4d4eded4b2fc/Lib/enum.py#L548). This ultimately leads us to [the error](https://github.com/python/cpython/blob/a6b889e99d5a8cc63924efacda9e4d4eded4b2fc/Lib/enum.py#L560) the users reported.

Together with `@dataclasses.dataclass`, the functionality that we want is trivial to implement. Re-using the example from above:

```python
import pickle
from copy import deepcopy

from torchvision.models._api import TransformsFactory


def foo():
    pass


bar = TransformsFactory(foo)
assert bar == deepcopy(bar)
assert bar == pickle.loads(pickle.dumps(bar))
```

Another option that we could do is what OP describes in #7099. Basically we need to implement a custom `__eq__` on the `Weights` class that is able to handle the `partial`. I found this a little more brittle, since we don't enforce anything about the callable and thus specializing the comparison to `partial` might lead to other errors. Note that in the `TransformsFactory` implementation, we don't need to touch this behavior at all and it is handled by `@dataclass` automatically.

ToDo:

- [ ] expand new architecture to all weights
- [ ] write tests that make sure both deepcopy and serialization works and gives the expected results

### CODE REVIEW

File: torchvision/models/_api.py [hunk 1]
```diff
@@ -1,19 +1,52 @@
...
+    "Weights",
+    "get_model",
+    "get_model_builder",
+    "get_model_weights",
+    "get_weight",
+    "list_models",
+]
+
+
+@dataclass(init=False)
```
Comment (NicolasHug) [inline torchvision/models/_api.py:29]: [Docs](https://docs.python.org/3/library/dataclasses.html#module-contents) say

> If the class already defines __init__(), this parameter is ignored.

so we may not need it?

```suggestion
@dataclass
```

File: torchvision/models/_api.py [hunk 2]
```diff
@@ -1,19 +1,52 @@
...
+class TransformsFactory:
+    fn: Callable[..., nn.Module]
+    args: Tuple[Any, ...]
+    kwargs: Dict[str, Any]
+
+    def __init__(self, fn: Callable[..., nn.Module], *args: Any, **kwargs: Any) -> None:
+        self.fn = fn
+        self.args = args
+        self.kwargs = kwargs
```
Comment (NicolasHug) [inline torchvision/models/_api.py:38]: Wouldn't the dataclass automatically generate this `__init__()`?

File: torchvision/models/_api.py [hunk 3]
```diff
@@ -1,19 +1,52 @@
...
 from torch import nn
 
 from torchvision._utils import StrEnum
 
 from .._internally_replaced_utils import load_state_dict_from_url
 
-
-__all__ = ["WeightsEnum", "Weights", "get_model", "get_model_builder", "get_model_weights", "get_weight", "list_models"]
+__all__ = [
+    "TransformsFactory",
```
Comment (NicolasHug) [inline torchvision/models/_api.py:18]: I assume we do not need this to be public? And if it's not public, does it need to be in `__all__`?

**Review by NicolasHug — COMMENTED** [review]
Thanks a lot Philip for looking into this,

The addition new `TransformsFactory` class looks like a reasonable solution. I just have a few questions below

> I found this a little more brittle, since we don't enforce anything about the callable and thus specializing the comparison to partial might lead to other errors

Are you saying that the `__eq__` proposed in the other issue *assumes* that the `transforms` fields are implemented as `partial`, and that this assumption isn't enforced and thus may break? (I agree)

Does the metaclass solution still work with torchscript?

File: torchvision/models/_api.py (hunk 1 above)
Comment (pmeier) [reply torchvision/models/_api.py:29]: Good catch!

File: torchvision/models/_api.py (hunk 2 above)
Comment (pmeier) [reply torchvision/models/_api.py:38]: Nope, it would create

```python
def __init__(self, fn, args, kwargs):
    ...
```

Note the missing tuple and dict unpacking. I wanted this to make `TransformsFactory` as close to `functools.partial` as possible. Otherwise, we need to do something like `TransformsFactory(ImageClassification, (), dict(crop_size=[224]))` rather than `TransformsFactory(ImageClassification, crop_size=[224])`

File: torchvision/models/_api.py (hunk 3 above)
Comment (pmeier) [reply torchvision/models/_api.py:18]: Right now we don't. But with #7094 in mind, this is probably a useful helper for users as well. No strong opinion though.

**Comment by pmeier** [comment]
> Are you saying that the `__eq__` proposed in the other issue _assumes_ that the `transforms` fields are implemented as `partial`, and that this assumption isn't enforced and thus may break? (I agree)

The proposal looks like this:

```python
if not isinstance(self.transforms, functools.partial) or not isinstance(other.transforms, functools.partial):
    return self.transforms == other.transforms
return all(
    getattr(self.transforms, a) == getattr(other.transforms, a)
    for a in ["func", "args", "keywords", "__dict__"]
)
```

So we have a special treatment for `functools.partial` and everything else will be handled by a regular `==`. So no, it doesn't assume that the attribute is a partial (I think we already have cases where it is not), but special cases it. The reason I somewhat dislike this is that there are potentially more callable types that need special handling. Of the top of my mind, what if someone would use `lambda: ImageClassification(crop_size=[224])`.

This ties a little bit into https://github.com/pytorch/vision/pull/7107#discussion_r1073425197. If we don't want a public helper, we are probably better off with the user suggestion since it is a best effort approach for arbitrary callables.

**Comment by NicolasHug** [comment]
Thanks for the details Philip.

Perhaps this is something @adamjstewart and @nilsleh can give their opinion on, since they're the ones who are going to be using those Weight outside of torchvision.

For some context: we're trying to address 2 different bugs with the `WeightEnum`: pickle-ability https://github.com/pytorch/vision/issues/7099 and support for `deepcopy` (and Lightning, apparently) https://github.com/pytorch/vision/issues/6871#issuecomment-1296916472.

We're thinking of 2 options:

- We override `__eq__` of `Weights` in torchvision: you have nothing to do in torchGeo, but we can only guarantee that the implem of `__eq__`  supports torchvision's needs: i.e. right now our  `transforms` field is implemented as `partial` so that's what we support.  Should you rely on something different (e.g. a `lambda` or something more esoteric), it may or may not break. If it breaks, we can address it in torchvision when you let us know, or you can also fix it in torchgeo by replacing `__eq__`, but it's of course a bit brittle. 
- We provide a public `TransformsFactory` helper (name TBD) which you would have to use when defining each of the transforms fields (https://github.com/pytorch/vision/pull/7107/files#diff-6aff9cc95a81d1947b2e5fed0363017759e6e54a6a4c047394a3ab61ffbc35d4R359)

Personally, I have a slight pref for overriding the `__eq__` because it avoids updating all the `transforms` fields, and it keeps the user-facing API simpler. But I'm interested in what @adamjstewart and @nilsleh might think

**Comment by adamjstewart** [comment]
Former is probably sufficient for us. Our current weights have no `transforms` field, we just use `nn.Identity`. Unfortunately, different datasets use different normalizations (0–1, 0–2^8, 0–2^16, etc.) so our users will have to figure out the appropriate normalizations themselves.

**Comment by NicolasHug** [comment]
Thanks for your input @adamjstewart . 

> Our current weights have no transforms field

Does that mean you wouldn't need the `Weight` class to be public, only the `WeightEnum` one?

@pmeier are you OK to go for `__eq__` or do you still have some concerns?

**Comment by pmeier** [comment]
I've pushed a version with a custom `Weights.__eq__`. @NicolasHug PTAL. If you are ok with the design, I'm going to write and fix tests.

**Comment by NicolasHug** [comment]
>  @NicolasHug PTAL. If you are ok with the design, I'm going to write and fix tests.

Yes this looks very clean, thank you! We may also want to add a comment in `__eq__` about why this is all needed. We may not need to go too deep in the details but at least explain that this is needed to handle `deepcopy` and `pickle`

**Review by NicolasHug — APPROVED** [review]
Thanks Philip, some minor comments / nits but LGTM anyway

File: torchvision/models/_api.py [hunk 4]
```diff
@@ -37,6 +38,30 @@ class Weights:
     transforms: Callable
     meta: Dict[str, Any]
 
+    def __eq__(self, other: Any) -> bool:
+        # We need this custom implementation for correct deep-copy and deserialization behavior.
+        # TL;DR: After the definition of an enum, creating a new instance, i.e. by deep-copying or deserializing it,
+        # involves an equality check against the defined members. Unfortunately,
```
[the remark this answers is not in the document]
Comment (pmeier) [reply torchvision/models/_api.py:44]: This is not always true:

https://github.com/pytorch/vision/blob/93df9a50885d0345e31bba691576c83d5cee7737/torchvision/models/detection/faster_rcnn.py#L381

Still, I'll update the comment to reflect that this is often the case.

File: test/test_extended_models.py [hunk 5]
```diff
@@ -73,10 +74,24 @@ def test_get_model_weights(name, weight):
     ],
 )
 def test_weights_copyable(copy_fn, name):
-    model_weights = models.get_model_weights(name)
-    for weights in list(model_weights):
-        copied_weights = copy_fn(weights)
-        assert copied_weights is weights
+    for weights in list(models.get_model_weights(name)):
+        assert copy_fn(weights) == weights
```
[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_extended_models.py:78]: > Isn't `copy_fn(weights) is weights` still True?

Yes.

> Maybe we should still keep it in the tests, and acknowledge in a comment that this is a somewhat surprising (but builtin) behaviour of enum fields?

I think we should ask whether we want to enforce it or not. I'm ok doing this to align with the builtin enums, but I think both an equality or identity check is fine. Your choice.

In any case, we should use the same for the serialization test.

[the remark this answers is not in the document]
Comment (NicolasHug) [reply test/test_extended_models.py:78]: I'd be in favour of adding the `is` check in addition to the `==` check. Not because this is something we want to enforce (I don't think we should, hence the suggested comments), but mostly as documentation in case similar issues pop-up in the future.

File: test/test_extended_models.py [hunk 6]
```diff
@@ -73,10 +74,30 @@ def test_get_model_weights(name, weight):
     ],
 )
 def test_weights_copyable(copy_fn, name):
-    model_weights = models.get_model_weights(name)
-    for weights in list(model_weights):
-        copied_weights = copy_fn(weights)
-        assert copied_weights is weights
+    for weights in list(models.get_model_weights(name)):
+        # It is somewhat surprising that (deep-)copying is an identity operation here,
+        # but this is the default behavior of enums.
```
Comment (NicolasHug) [inline test/test_extended_models.py:79]: Still LGTM but here and below could we just

- link to the official docs https://docs.python.org/3/howto/enum.html#enum-members-aka-instances
- clarify that this isn't a behaviour that we want to enforce and that `==` would be fine and even preferable, should we change it in the future

**Review by NicolasHug — APPROVED** [review]
Still LGTM! thanks
